tfsdk/serve_import.go

```diff
@@ -0,0 +1,128 @@
+package tfsdk
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+// importedResource represents a resource that was imported.
+//
+// This type is not exported as the framework import implementation is
+// currently designed for the most common use case of single resource import.
+type importedResource struct {
+	Private  []byte
+	State    State
+	TypeName string
+}
+
+func (r importedResource) toTfprotov6(ctx context.Context) (*tfprotov6.ImportedResource, diag.Diagnostics) {
+	var diags diag.Diagnostics
+	irProto6 := &tfprotov6.ImportedResource{
+		Private:  r.Private,
+		TypeName: r.TypeName,
+	}
+
+	stateProto6, err := tfprotov6.NewDynamicValue(r.State.Schema.TerraformType(ctx), r.State.Raw)
+
+	if err != nil {
+		diags.AddError(
+			"Error converting imported resource response",
+			"An unexpected error was encountered when converting the imported resource response to a usable type. This is always a problem with the provider. Please give the following information to the provider developer:\n\n"+err.Error(),
+		)
+		return nil, diags
+	}
+
+	irProto6.State = &stateProto6
+
+	return irProto6, diags
+}
+
+// importResourceStateResponse is a thin abstraction to allow native Diagnostics usage
+type importResourceStateResponse struct {
+	Diagnostics       diag.Diagnostics
+	ImportedResources []importedResource
+}
+
+func (r importResourceStateResponse) toTfprotov6(ctx context.Context) *tfprotov6.ImportResourceStateResponse {
+	resp := &tfprotov6.ImportResourceStateResponse{
+		Diagnostics: r.Diagnostics.ToTfprotov6Diagnostics(),
+	}
+
+	for _, ir := range r.ImportedResources {
+		irProto6, diags := ir.toTfprotov6(ctx)
+		resp.Diagnostics = append(resp.Diagnostics, diags.ToTfprotov6Diagnostics()...)
+		resp.ImportedResources = append(resp.ImportedResources, irProto6)
+	}
+
+	return resp
+}
+
+func (s *server) importResourceState(ctx context.Context, req *tfprotov6.ImportResourceStateRequest, resp *importResourceStateResponse) {
+	resourceType, diags := s.getResourceType(ctx, req.TypeName)
+	resp.Diagnostics.Append(diags...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resourceSchema, diags := resourceType.GetSchema(ctx)
+	resp.Diagnostics.Append(diags...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resource, diags := resourceType.NewResource(ctx, s.p)
+	resp.Diagnostics.Append(diags...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	emptyState := tftypes.NewValue(resourceSchema.TerraformType(ctx), nil)
```

Comment:
A thing worth thinking about here:

We need a resolution to #148 to make this work the way we want.

Another option is to have it do something like:

```go
vals := map[string]tftypes.Value{}
typ := resourceSchema.TerraformType(ctx)
for name, t := range typ.(tftypes.Object).AttributeTypes {
  vals[name] = tftypes.NewValue(t, nil)
}
emptyState := tftypes.NewValue(typ, vals)
```

This matches the more-expected (imho) default state, which is an object in state with nothing filled in, instead of nothing in state.

However, this is weird during failure cases. I'm not clear on whether Terraform will persist the object to state or write nothing to state in the face of an error diagnostic. If the former, we should default to an empty state and just rely on #148. If the latter, we should confirm that will continue to be the case, and maybe defaulting to the empty object (instead of the null object) is more expected? I dunno. I don't know that it makes a huge difference, just wanted to call it out here.

Reply:
The current logic in Terraform CLI is to return early on ImportResourceState error diagnostics, before saving state information:

https://github.com/hashicorp/terraform/blob/2afa0a5e75d76de7e47157114c579b3b1bff994f/internal/terraform/transform_import_state.go#L141-L154

Reply:
🤔 how do we feel about just leaving it and waiting for feedback?

Reply:
We're going to merge this in as-is for now and iterate if necessary.

If we do wind up switching the implementation here in the future -- maybe creating the above code snippet as a method such as `(Schema).newState()` would be good to try and prevent "empty" object problems here and in any other potential places that might need it. That might be something we could export if/when multiple resource import support makes sense.